### ISSUE TITLE
Common mode estimation is different at each computation

### ISSUE THREAD

**Issue by angeliquemarck** [tobiscode/disstans#12]
Hi, 

I was running Disstans following the steps in Tutorial 3 and Example 1 to estimate trajectory models on my data. While performing the cleaning steps and removing the common mode, I noticed that the common mode estimate is not stable for my dataset.
The spatial pattern often flips between runs, and the temporal pattern also varies, with residuals between two runs up to about 1 mm in the horizontal components and 2 mm in the vertical.
When I run the same procedure on the synthetic data from Tutorial 3, I don’t encounter this problem: the spatial pattern is consistent, and the difference between temporal components is zero. However, I observe the same instability when using the Example 1 dataset, following the same processing steps:
```
net = ds.Network("LVC")
for _, row in df_filtered.iterrows():
    # get name and location of station
    name = row["Sta"]
    loc = [row["Lat"], row["Long"], row["Hgt"]]
    # make a timeseries object to check availability metric
    tspath = f"{gnss_dir}/{name}.tenv3"
    loaded_ts = ds.timeseries.UNRTimeseries(tspath)
    loaded_ts.cut(t_max="2022-01-01")
    # make a station and add the timeseries only if two quality metrics are met
    if loaded_ts.shape[0] > 0:
        if (loaded_ts.reliability > 0.5) and (loaded_ts.length > pd.Timedelta(365, "D")):
            net[name] = ds.Station(name, loc)
            net[name]["raw"] = loaded_ts
net.gui(station="CASA", gui_kw_args={"wmts_show": True, "wmts_alpha": 0.5})
# compute reference
net.call_func_ts_return("median", ts_in="raw", ts_out="raw_filt", kernel_size=7)
# remove outliers
net.call_func_no_return("clean", ts_in="raw", reference="raw_filt", ts_out="raw_clean")
# get the residual for each station
net.math("raw_filt_res", "raw_clean", "-", "raw_filt")
# calculate common mode
net.graphical_cme(ts_in="raw_filt_res", ts_out="CME", annotate_stations=True, save=False, save_kw_args={"dpi":300,"format":"png"},method="ica",detrend=False,rng=rng)
net.graphical_cme(ts_in="raw_filt_res", ts_out="CME_2", annotate_stations=True, save=False, save_kw_args={"dpi":300,"format":"png"},method="ica",detrend=False,rng=rng)
net.math("CME_diff","CME_2","-","CME")
```
I'm using:
`rng = np.random.default_rng(0)`
Disstans version 2.2, python 3.13.7.

I’m attaching:
- the CASA time series displayed by the GUI (which looks the same as to the one in the example),
- the two spatial patterns output by graphical_cme
- two examples of CME_diff time series for stations CASA and HOTK.

I was wondering whether this behaviour is expected, or if it might indicate an issue with my setup or processing steps ?

Best regards, 
Angélique 

<img width="651" height="491" alt="Image" src="https://github.com/user-attachments/assets/b6f79daf-0c04-48ee-8f69-484a67788e75" />
<img width="654" height="495" alt="Image" src="https://github.com/user-attachments/assets/6f6d10f9-4de6-48af-a590-8b3b17dcdc05" />
<img width="580" height="445" alt="Image" src="https://github.com/user-attachments/assets/dbd5543c-0f5b-480f-a9fd-75e3529ee12d" />
<img width="989" height="690" alt="Image" src="https://github.com/user-attachments/assets/1c830132-d86e-44ac-8a32-fcaf649b886d" />
<img width="998" height="691" alt="Image" src="https://github.com/user-attachments/assets/448513e0-d6b0-4435-902e-402579293f05" />

**Comment by tobiscode** [comment]
Hi,

you're right that if you always use the same seed for the RNG, that the common mode estimation should always be the same. However, at least in the code you shared, you don't reset the RNG before or in between the calls to `graphical_cme`, so I'm not surprised you get different results. Even though it's the same object (`rng`), each successive use of it will be dependent on the previous one (i.e., the sequence of random numbers it generates will continue, and not repeat). Can you test whether this would give you the same results?

```python
# calculate common mode
rng = np.random.default_rng(0)
net.graphical_cme(ts_in="raw_filt_res", ts_out="CME", annotate_stations=True, save=False, save_kw_args={"dpi":300,"format":"png"},method="ica",detrend=False,rng=rng)
rng = np.random.default_rng(0)
net.graphical_cme(ts_in="raw_filt_res", ts_out="CME_2", annotate_stations=True, save=False, save_kw_args={"dpi":300,"format":"png"},method="ica",detrend=False,rng=rng)
```

(I only put `rng = np.random.default_rng(0)` once in the top of the scripts (and not in between) such that the overall result of each script will be the same.)

Cheers

**Comment by angeliquemarck** [comment]
Thanks for your quick response. 

The spatial patterns can still appear flipped between runs with the graphical_cme but the common-mode time series output are always the same with the reset.

Cheers,

**Comment by tobiscode** [comment]
Hi,

glad that it helped - it still shouldn't flip, so I looked up the PCA and ICA implementations and sure enough, they also rely on RNGs in most cases. I currently don't pass the RNG used by DISSTANS to the PCA and ICA implementation, so that is probably the reason it still flips. You can try using PCA, then depending on the input data, it won't use an RNG (see the description for the `svd_solver` parameter [here](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)).

I'll see if I'll update the code in the future to make the PCA and ICA always reproducible.

Cheers

**Comment by angeliquemarck** [comment]
Thanks!

A last question. Would you trust the outputted CME when it is different depending on the rng (on synthetics it is not) or does it mean the estimation is currently not stable ?

Cheers,

**Comment by tobiscode** [comment]
TL;DR: Yes, I would still trust it.

First, it's an estimation of a noise term, based on heuristics that may contain randomness (PCA/ICA), so the result naturally could have some randomness associated with it.

Second, if you look into the estimated CME in the cases where the spatial pattern still flips, you will probably find that the sign of the spatial and temporal components just flip. Combined, they should still produce the same result at any given station (hence why the CME timeseries at each station now look the same). To see that, what you can try is that after estimating the CME twice, without resetting the RNG, to flip the sign of the the temporal component such that the spatial component all point into the same quadrant (e.g., all north-east). Then, the temporal components should also mostly match. (Disclaimer: this check is from my memory, I haven't tried this recently and don't have any code.)

Cheers
